extmod/modselect.c

```diff
@@ -26,62 +26,223 @@
  */
 
 #include "py/mpconfig.h"
-#if MICROPY_PY_SELECT
-
-#include <stdio.h>
-
 #include "py/runtime.h"
 #include "py/obj.h"
 #include "py/objlist.h"
 #include "py/stream.h"
 #include "py/mperrno.h"
 #include "py/mphal.h"
 
+#if MICROPY_PY_SELECT
+
+#if MICROPY_PY_SELECT_SELECT && MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+#error select.select is not supported with MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+#endif
+
+#if MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+#include <poll.h>
+#endif
+
 // Flags for poll()
 #define FLAG_ONESHOT (1)
 
+// A single pollable object.
 typedef struct _poll_obj_t {
     mp_obj_t obj;
     mp_uint_t (*ioctl)(mp_obj_t obj, mp_uint_t request, uintptr_t arg, int *errcode);
+    #if MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+    struct pollfd *pollfd;
+    uint16_t nonsys_flags;
+    uint16_t nonsys_flags_ret;
+    #else
     mp_uint_t flags;
     mp_uint_t flags_ret;
+    #endif
 } poll_obj_t;
 
-STATIC void poll_map_add(mp_map_t *poll_map, const mp_obj_t *obj, mp_uint_t obj_len, mp_uint_t flags, bool or_flags) {
+// A group of pollable objects.
```

Comment:
Initially I was confused by what a "poll group" was, I think I expected maybe you'd group-by the poll-style (i.e. fd or no-fd). But there's never more than one group, rather it's a convenient way of bundling the map with the list of pollfds (and metadata for that list).

(I think "list" or "set" is a better term than "group", because "group" implies they're grouped by something).

Reply:
Renamed `poll_group_t` to `poll_set_t`.